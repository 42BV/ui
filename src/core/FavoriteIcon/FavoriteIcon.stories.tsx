import { useState } from 'react';
import { storiesOf } from '@storybook/react';

import { FavoriteIcon } from './FavoriteIcon';

storiesOf('core/FavoriteIcon', module)
  .addParameters({ component: FavoriteIcon })
  .add('default', () => {
    const [favorite, setFavorite] = useState(false);
    return (
      <div className="text-center">
        <FavoriteIcon
          onChange={() => setFavorite(!favorite)}
          value={favorite}
        />
      </div>
    );
  })
  .add('hover color', () => {
    const [favorite, setFavorite] = useState(false);
    return (
      <div className="text-center">
        <FavoriteIcon
          onChange={() => setFavorite(!favorite)}
          value={favorite}
          hoverColor="danger"
        />
      </div>
    );
  });
