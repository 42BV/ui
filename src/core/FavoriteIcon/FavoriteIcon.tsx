import React from 'react';
import Icon from '../Icon/Icon';
import { useHover } from '../../hooks/useHover/useHover';
import { Color } from '../types';

type FavoriteIconProps = {
  /**
   * The value that the form element currently has.
   */
  value: boolean;

  /**
   * Callback for when the form element changes.
   */
  onChange: (value: boolean) => void;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * Optionally the size of the icon in pixels.
   *
   * Defaults to `24px`
   */
  size?: number;

  /**
   * Optionally the color of the icon when value is false
   * and the icon is not hovered.
   *
   * Defaults to secondary.
   */
  color?: Color;

  /**
   * Optionally the color of the icon when value is true
   * or the icon is hovered.
   *
   * Defaults to primary.
   */
  activeColor?: Color;
};

/**
 * FavoriteIcon displays a star icon to show the user
 * whether an item is favorite or not. The component
 * includes toggle functionality.
 */
export function FavoriteIcon({
  value,
  onChange,
  className = '',
  size,
  color = 'secondary',
  activeColor = 'primary'
}: FavoriteIconProps) {
  const [hover, hoverEvents] = useHover();

  return (
    <div {...hoverEvents}>
      <Icon
        color={value || hover ? activeColor : color}
        className={className}
        icon={value ? 'star' : 'star_border'}
        onClick={() => onChange(!value)}
        size={size}
      />
    </div>
  );
}
