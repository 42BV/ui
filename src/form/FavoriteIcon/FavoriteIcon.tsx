import React from 'react';
import Icon from '../../core/Icon/Icon';
import { useHover } from '../../hooks/useHover/useHover';

type FavoriteIconProps = {
  /**
   * The value that the form element currently has.
   */
  value: boolean;

  /**
   * Callback for when the form element changes.
   */
  onChange: (value: boolean) => void;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
};

/**
 * FavoriteIcon displays a star icon to show the user
 * whether an item is favorite or not. The component
 * includes toggle functionality.
 */
export function FavoriteIcon({
  value,
  onChange,
  className = ''
}: FavoriteIconProps) {
  const [hover, hoverEvents] = useHover();

  return (
    <div {...hoverEvents}>
      <Icon
        color={value || hover ? 'primary' : 'secondary'}
        className={className}
        icon={value ? 'star' : 'star_border'}
        onClick={() => onChange(!value)}
      />
    </div>
  );
}
