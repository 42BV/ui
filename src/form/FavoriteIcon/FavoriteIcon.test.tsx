import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { FavoriteIcon } from './FavoriteIcon';

describe('Component: FavoriteIcon', () => {
  function setup({ value }: { value: boolean }) {
    const onChangeSpy = jest.fn();

    const favoriteIcon = shallow(
      <FavoriteIcon onChange={onChangeSpy} value={value} />
    );

    return { favoriteIcon, onChangeSpy };
  }

  describe('ui', () => {
    test('favorite', () => {
      const { favoriteIcon } = setup({ value: true });

      expect(toJson(favoriteIcon)).toMatchSnapshot(
        'Component: FavoriteIcon => ui => favorite'
      );
    });

    test('not favorite', () => {
      const { favoriteIcon } = setup({ value: false });

      expect(toJson(favoriteIcon)).toMatchSnapshot(
        'Component: FavoriteIcon => ui => not favorite'
      );
    });
  });

  describe('events', () => {
    it('should when value is false and the icon is clicked call onChange with true', () => {
      const { favoriteIcon, onChangeSpy } = setup({ value: false });

      favoriteIcon.find('Icon').simulate('click');

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith(true);
    });

    it('should when value is true and the icon is clicked call onChange with false', () => {
      const { favoriteIcon, onChangeSpy } = setup({ value: true });

      favoriteIcon.find('Icon').simulate('click');

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith(false);
    });
  });
});
