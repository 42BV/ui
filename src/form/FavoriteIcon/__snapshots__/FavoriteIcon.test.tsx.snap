// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: FavoriteIcon ui favorite: Component: FavoriteIcon => ui => favorite 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Icon
    className=""
    color="primary"
    icon="star"
    onClick={[Function]}
  />
</div>
`;

exports[`Component: FavoriteIcon ui not favorite: Component: FavoriteIcon => ui => not favorite 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Icon
    className=""
    color="secondary"
    icon="star_border"
    onClick={[Function]}
  />
</div>
`;
